README.md
# Reference implementation of SynDiffix

## Purpose

This is the reference implementation of SynDiffix. It is two primary roles:
* To test the properties and performance of SynDiffix.
* To serve as the core description of the technology from which other implementations and integrations can be built.

## Usage

From inside the `src\SynDiffix` folder execute:

```
dotnet run file.csv --columns column0:r column1:i ...
```

`<file.csv>` - Specifies the input CSV file.

`--columns [<string>...]` - The selection of columns to synthesize and their types in the format `name:x`, where `x` is one of the following:
`b`-boolean, `i`-integer, `r`-real, `t`-timestamp, `s`-string.

### Python wrapper for ML

The Python wrapper `syndiffix.py` uses [scikit-learn](https://scikit-learn.org/) to find the most relevant ML features for a target column.

```
python syndiffix.py <input-path> <output-path> --dev --ml_target=<ml-target-column>
```

Run `syndiffix.py --help` for usage guide.

### Optional arguments

`--output <string>, -o <string>` - Path for CSV output. If not specified, result is printed to stdout.

`--aidcolumns [<string>...]` - List of entity ID columns used for anonymization. If not specified,
assumes each row represents a distinct protected entity.

`--verbose` - Display extra output for debugging purposes.

#### Anonymization parameters

`--lcf-low-threshold, -lcf <int>` - Low threshold for the low-count filter.

`--range-low-threshold, -range <int>` - Low threshold for a range bucket.

`--singularity-low-threshold, -sing <int>` - Low threshold for a singularity bucket.

`--threshold-sd, -thresh_sd <double>` - Threshold SD for LCF/range/singularity decisions.

`--outlier-count <int> <int>` - Outlier count interval.

`--top-count <int> <int>` - Top values count interval.

`--layer-noise-sd, -noise_sd <double>` - Count SD for each noise layer.

#### Clustering parameters

`--no-clustering` - Disables column clustering.

`--clustering-maincolumn <string>` - Column to be prioritized in clusters for better ML accuracy.

`--clustering-mainfeatures [<string>...]` - Best features of main column. When specified, this overrides the default clustering algorithm.

`--clustering-samplesize <int>` - Table sample size when measuring dependence.

`--clustering-maxweight <double>` - Maximum cluster weight. Default value of 15.0 is approximately 4-7 columns, depending on their variance.

`--clustering-thresh-merge <double>` - Dependence threshold for combining columns in a cluster.

#### Precision parameters

`--precision-limit-row-fraction <int>` - Tree nodes are allowed to split if `node_num_rows >= table_num_rows/row_fraction`.

`--precision-limit-depth-threshold <int>` - Tree depth threshold below which the `row-fraction` check is not applied.

## Usage guide

The fewer columns that are included with the `--columns` parameter, the more accurate the resulting synthetic data. If for instance one is only interested in column pairs, then it is better to generate a different synthetic dataset for each pair than to take the pairs from one complete synthetic dataset.

If the synthetic dataset is for the purpose of building an ML model, and the target column of the model is known, then setting `--clustering-maincolumn` to the target column will produce synthetic data that leads to a better ML model.

Running syndiffix through the [Python wrapper](#python-wrapper-for-ml) will automatically find the best ML features for the target column.
This results in better ML quality compared to the univariate measures that syndiffix uses internally for feature detection.

If there is more than one row per protected entity (i.e. per individual), then the `--aidcolumns` parameter must be set.

All other parameters should normally be kept at their defaults.

## Step-by-step description of the algorithm

All of the following steps, whenever a `count` is mentioned, refer to the `anonymized (noisy) count`
(the `true count` is never used directly).

### Casting to real

Before the forest building is conducted, preliminary casting of all values to reals is done:

- **null** - see [Forest building](#forest-building)
- **boolean** - false becomes 0.0, true becomes 1.0
- **integer** - natural cast to real
- **real** - kept intact (TODO: document behavior for non-finite reals, once it's implemented)
- **string** - a sorted array with all of the distinct values is first created, then each value gets assigned
  its position in the array
- **timestamp** - becomes seconds since 1800-01-01 00:00:00 UTC

### Forest building

There are two types of nodes: branches and leaves. Both kinds store an array of ranges, an array of subnodes, and a stub flag.
A branch has `2^tree_dimensions` children, some of which can be missing. A leaf stores the list of rows matching its ranges.

- For each row:
  - For each column:
    - Compute the range of the values space.
- For each column:
  - Map nulls to a value outside twice the actual min or max boundary,
    to prevent leaking low-count nulls into the bottom or top values range.
  - Snap the range size to the next power-of-two.
  - Align the range boundaries to the snapped size.
- For k in 1..length(columns):
  - For each combination of k columns:
    - Build the tree root for the column combination.
    - For each row:
      - Find target leaf node. If one doesn't exist:
        - Create a new leaf node by halfing the ranges of the parent.
        - Fill the subnodes list from the lower dimensional trees.
          - The subnodes are the children of parent's subnodes with the same corresponding ranges as the current node.
        - Determine if the node is a stub.
          - A node is a stub if all of its subnodes either:
            - are missing;
            - fail the range / singularity threshold;
            - are also stubs themselves.
      - Add the row to its matching leaf node.
      - If target node is not a singularity, passes LCF, is not a stub, and (below depth threshold or over row fraction):
        - Convert the target leaf node to a branch node.
        - Add the previously stored rows to the newly created node.
    - If k = 1:
      - Push the root down as long as one of its children fails LCF.
      - Add the rows from the previously dropped low-count nodes to the remaining tree.
      - Update the snapped range for the corresponding dimension.

### Bucket harvesting

A bucket has an array of ranges (one for each dimension), and a noisy row count.
When creating a bucket from a node, use the node's singularity ranges at its singularity dimensions
and its snapped ranges elsewhere.

- Do a bottom-up traversal of the forest's root. At each node:
  - If the node is a leaf:
    - If the node fails LCF:
      - Return nothing.
    - If the node is a singularity:
      - Return a bucket from node's single value.
    - If in 1-dimensional node:
      - Return a bucket from node's ranges.
    - Else:
      - Return the refined buckets from node's ranges and subnodes.
  - Else:
    - Gather all buckets from child nodes.
    - If there are fewer rows in the children's buckets than half of the current node's rows:
      - If in 1-dimensional node:
        - Return a bucket from node's ranges.
      - Else:
        - Generate `current.count - children.count` refined buckets from node's ranges and subnodes.
        - Append the refined buckets to the previously harvested buckets.
        - Return the resulting buckets.
    - Else:
      - Adjust the counts of the buckets so that their sum matches the count of the current node.
      - Return the resulting buckets.

#### Bucket refinement

- Gather the sub-dimensional buckets. For each possible subnode:
  - If the subnode's combination corresponds to a singularity:
    - Return a bucket with singularity ranges for it.
  - If the subnode exists:
    - Harvest and return the buckets for it.
  - Else:
    - Return an empty list.
- If insufficient buckets for a subnode:
  - Return the input bucket.
- Compute the possible cumulative ranges for each subnode and dimension.
- Compute the smallest ranges for all dimensions.
- Compact smallest ranges by excluding any low-count half-ranges.
- Gather the possible per-dimension ranges which are inside the smallest ranges.
- Drop the sub-dimensional buckets which are outside the smallest ranges.
- If insufficient ranges:
  - Return the input bucket.
- Else:
  - Generate new buckets by matching the remaining possible ranges of a column with
    their corresponding sub-dimensional buckets using a deterministic random shuffle.

### Microdata generation

- For each harvested bucket:
  - For 1..bucket.count:
    - For each range in the bucket:
      - Generate a random float value inside the range.
      - Cast the generated value back to the corresponding column type.
        - In the case of **string** columns:
          - If the range is a singularity:
            - Return the exact value.
          - Else:
            - Return the common prefix of the range, plus `*`, plus a random integer from inside the range.
    - Return a synthetic row from the generated values.

## More information

A description of SynDiffix, its privacy properties, and its performance measured over a variety of datasets can be found at [https://github.com/diffix/syndiffix/wiki/SynDiffix:-Overview](https://github.com/diffix/syndiffix/wiki/SynDiffix:-Overview).

CHANGELOG.md
## Changelog

### Version 1.1.0

- Added Python wrapper for auto-detecting column types and main features for ML.
- Lowered default thresholds for range and singularity nodes and raised default tree depth limit.
- Improved clustering algorithm for main column.
- Added `--output` (`-o`) CLI argument to directly save the CSV file to disk.
- Added `--clustering-mainfeatures <features>` CLI argument to specify main column's ML features.
- Added `--clusters <clusters>` CLI argument which allows defining clusters manually.

### Version 1.0.2

- Fixed a bug in the computation of low-count/range/singularity thresholds' mean.

### Version 1.0.1

- Fixed conversion bug when generating boolean microdata.


features.py
import pandas as pd
import numpy as np
from sklearn.compose import ColumnTransformer
from sklearn.feature_selection import RFECV, VarianceThreshold
from sklearn.preprocessing import FunctionTransformer, OneHotEncoder, OrdinalEncoder, RobustScaler
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor


def is_classification(column):
    return column.dtype == 'object' or column.nunique() <= 10


def get_feature_types(df):
    text_features = []
    continuous_features = []
    categorical_features = []

    for colname in df.columns:
        column = df[colname]
        nunique = column.nunique()
        if nunique <= 10:
            categorical_features.append(colname)
        elif column.dtype == 'object':
            text_features.append(colname)
        else:
            continuous_features.append(colname)

    return text_features, continuous_features, categorical_features


def one_hot_feature_names(encoder, categorical_features):
    cats = [
        encoder._compute_transformed_categories(i)
        for i, _ in enumerate(encoder.categories_)
    ]

    feature_names = []
    inverse_names = {}

    for i in range(len(cats)):
        category = categorical_features[i]
        names = [category + "$$" + str(label) for label in cats[i]]
        feature_names.extend(names)

        for name in names:
            inverse_names[name] = category

    return feature_names, inverse_names


def random_encoder(X):
    X = X.copy()

    for column in X.columns:
        column_text = X[column].apply(str)
        unique_values = column_text.unique()
        np.random.shuffle(unique_values)
        mapping = {value: index for index, value in enumerate(unique_values)}
        X[column] = column_text.map(mapping).astype(int)

    return X


def preprocess(df, one_hot_encode=True, variance_threshold=True):
    text_features, continuous_features, categorical_features = get_feature_types(df)
    ordinal_features = []

    if not one_hot_encode:
        ordinal_features = categorical_features
        categorical_features = []

    preprocessor = ColumnTransformer(
        transformers=[
            ('random_encode', FunctionTransformer(random_encoder), text_features),
            ('ord', OrdinalEncoder(), ordinal_features),
            ('num', RobustScaler(), continuous_features),
            ('cat', OneHotEncoder(sparse_output=False), categorical_features),
        ],
        sparse_threshold=0)

    result = preprocessor.fit_transform(df)

    inverse_lookup = {
        name: name for name in text_features + ordinal_features + continuous_features
    }

    categorical_names = []

    if len(categorical_features) > 0:
        categorical_names, categorical_inverse = one_hot_feature_names(
            preprocessor.named_transformers_['cat'],
            categorical_features
        )
        inverse_lookup.update(categorical_inverse)

    df_preprocessed = pd.DataFrame(
        result,
        columns=text_features + ordinal_features + continuous_features + categorical_names
    )

    if not variance_threshold:
        return df_preprocessed

    threshold = VarianceThreshold(0.00001)
    threshold.set_output(transform="pandas")

    df_filtered = threshold.fit_transform(df_preprocessed)

    return df_filtered, inverse_lookup


def convert_datetime_to_numeric(df):
    from pandas.errors import ParserError
    REFERENCE_DATE = pd.to_datetime('1800-01-01')

    for col_name, col_type in zip(df.columns, df.dtypes):
        if pd.api.types.is_datetime64_any_dtype(col_type):
            df[col_name] = pd.to_numeric(df[col_name] - REFERENCE_DATE).div(1e9).astype(float)

    return df


def split(df, column, one_hot_X):
    df.dropna(axis=0, inplace=True)
    convert_datetime_to_numeric(df)
    X, X_inv = preprocess(df.drop(column, axis=1), one_hot_encode=one_hot_X)
    y, _ = preprocess(df[[column]], one_hot_encode=False)
    return X, X_inv, y


def select_features_ml(df, column, one_hot_X=False):
    X, X_inv, y = split(df, column, one_hot_X)

    if y.shape[0] == 0 or y.shape[1] == 0:
        return {
            'valid': False,
            'features': [],
            'k': 0,
            'kFeatures': [],
            'cumulativeScore': [],
            'cumulativeScoreStd': [],
            'encoded': {
                'features': [],
                'k': 0,
                'kFeatures': [],
                'cumulativeScore': [],
                'cumulativeScoreStd': []
            }
        }

    if X.shape[1] == 1:
        return {
            'valid': False,
            'features': [X.columns[0]],
            'k': 1,
            'kFeatures': [X.columns[0]],
            'cumulativeScore': [0.0],
            'cumulativeScoreStd': [0.0],
            'encoded': {
                'features': [X.columns[0]],
                'k': 1,
                'kFeatures': [X.columns[0]],
                'cumulativeScore': [0.0],
                'cumulativeScoreStd': [0.0]
            }
        }

    if is_classification(y[column]):
        estimator = DecisionTreeClassifier()
    else:
        estimator = DecisionTreeRegressor()

    rfecv = RFECV(estimator=estimator)
    rfecv.fit(X, y)

    feature_ranks = rfecv.ranking_
    feature_names = X.columns.tolist()

    sorted_features = sorted(zip(feature_names, feature_ranks), key=lambda x: x[1])

    encoded_k = int(rfecv.n_features_)
    encoded_features = [name for name, _ in sorted_features]
    encoded_scores = rfecv.cv_results_["mean_test_score"].tolist()
    encoded_scores_std = rfecv.cv_results_["std_test_score"].tolist()

    decoded_k = 0
    decoded_features = []
    decoded_scores = []
    decoded_scores_std = []

    for i, feature in enumerate(encoded_features):
        decoded_feature = X_inv[feature]
        if decoded_feature in decoded_features:
            if decoded_features[-1] == decoded_feature:
                decoded_scores[-1] = encoded_scores[i]
                decoded_scores_std[-1] = encoded_scores_std[i]
        else:
            decoded_features.append(decoded_feature)
            decoded_scores.append(encoded_scores[i])
            decoded_scores_std.append(encoded_scores_std[i])

        if i == encoded_k - 1:
            decoded_k = len(decoded_features)

    return {
        'valid': True,
        'features': decoded_features,
        'k': decoded_k,
        'kFeatures': decoded_features[:decoded_k],
        'cumulativeScore': decoded_scores,
        'cumulativeScoreStd': decoded_scores_std,
        'encoded': {
            'features': encoded_features,
            'k': encoded_k,
            'kFeatures': encoded_features[:encoded_k],
            'cumulativeScore': encoded_scores,
            'cumulativeScoreStd': encoded_scores_std
        }
    }


syndiffix.py
import os
import platform
import subprocess
import shlex
import fire
import numpy as np
import pandas as pd
from sklearn.compose import ColumnTransformer
from sklearn.feature_selection import RFECV, VarianceThreshold
from sklearn.preprocessing import FunctionTransformer, OneHotEncoder, OrdinalEncoder, RobustScaler
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor

# ----------------------------------------------------------------
# Feature Selection
# ----------------------------------------------------------------


CATEGORY_THRESHOLD = 15


def is_categorical(column):
    return column.dtype == 'object' or column.nunique() <= CATEGORY_THRESHOLD


def get_feature_types(df):
    text_features = []
    continuous_features = []
    categorical_features = []

    for colname in df.columns:
        column = df[colname]
        nunique = column.nunique()
        if nunique <= CATEGORY_THRESHOLD:
            categorical_features.append(colname)
        elif column.dtype == 'object':
            text_features.append(colname)
        else:
            continuous_features.append(colname)

    return text_features, continuous_features, categorical_features


def one_hot_feature_names(encoder, categorical_features):
    cats = [
        encoder._compute_transformed_categories(i)
        for i, _ in enumerate(encoder.categories_)
    ]

    feature_names = []
    inverse_names = {}

    for i in range(len(cats)):
        category = categorical_features[i]
        names = [category + '$$' + str(label) for label in cats[i]]
        feature_names.extend(names)

        for name in names:
            inverse_names[name] = category

    return feature_names, inverse_names


def random_encoder(X):
    X = X.copy()

    for column in X.columns:
        column_text = X[column].apply(str)
        unique_values = column_text.unique()
        np.random.shuffle(unique_values)
        mapping = {value: index for index, value in enumerate(unique_values)}
        X[column] = column_text.map(mapping).astype(int)

    return X


def preprocess(df, one_hot_encode=True, variance_threshold=True):
    text_features, continuous_features, categorical_features = get_feature_types(df)
    ordinal_features = []

    if not one_hot_encode:
        ordinal_features = categorical_features
        categorical_features = []

    preprocessor = ColumnTransformer(
        transformers=[
            ('random_encode', FunctionTransformer(random_encoder), text_features),
            ('ord', OrdinalEncoder(), ordinal_features),
            ('num', RobustScaler(), continuous_features),
            ('cat', OneHotEncoder(sparse_output=False), categorical_features),
        ],
        sparse_threshold=0)

    result = preprocessor.fit_transform(df)

    inverse_lookup = {
        name: name for name in text_features + ordinal_features + continuous_features
    }

    categorical_names = []

    if len(categorical_features) > 0:
        categorical_names, categorical_inverse = one_hot_feature_names(
            preprocessor.named_transformers_['cat'],
            categorical_features
        )
        inverse_lookup.update(categorical_inverse)

    df_preprocessed = pd.DataFrame(
        result,
        columns=text_features + ordinal_features + continuous_features + categorical_names
    )

    if not variance_threshold:
        return df_preprocessed

    threshold = VarianceThreshold(0.00001)
    threshold.set_output(transform='pandas')

    df_filtered = threshold.fit_transform(df_preprocessed)

    return df_filtered, inverse_lookup


def convert_datetime_to_numeric(df):
    from pandas.errors import ParserError
    REFERENCE_DATE = pd.to_datetime('1800-01-01')

    for col_name, col_type in zip(df.columns, df.dtypes):
        if pd.api.types.is_datetime64_any_dtype(col_type):
            df[col_name] = pd.to_numeric(df[col_name] - REFERENCE_DATE).div(1e9).astype(float)

    return df


def split(df, column, one_hot_X):
    df.dropna(axis=0, inplace=True)
    convert_datetime_to_numeric(df)
    X, X_inv = preprocess(df.drop(column, axis=1), one_hot_encode=one_hot_X)
    y, _ = preprocess(df[[column]], one_hot_encode=False)
    return X, X_inv, y


def select_features_ml(df, column, one_hot_X=False):
    X, X_inv, y = split(df, column, one_hot_X)

    if y.shape[0] == 0 or y.shape[1] == 0:
        return {
            'valid': False,
            'features': [],
            'k': 0,
            'kFeatures': [],
            'cumulativeScore': [],
            'cumulativeScoreStd': [],
            'encoded': {
                'features': [],
                'k': 0,
                'kFeatures': [],
                'cumulativeScore': [],
                'cumulativeScoreStd': []
            }
        }

    if X.shape[1] == 1:
        return {
            'valid': False,
            'features': [X.columns[0]],
            'k': 1,
            'kFeatures': [X.columns[0]],
            'cumulativeScore': [0.0],
            'cumulativeScoreStd': [0.0],
            'encoded': {
                'features': [X.columns[0]],
                'k': 1,
                'kFeatures': [X.columns[0]],
                'cumulativeScore': [0.0],
                'cumulativeScoreStd': [0.0]
            }
        }

    if is_categorical(y[column]):
        estimator = DecisionTreeClassifier()
    else:
        estimator = DecisionTreeRegressor()

    rfecv = RFECV(estimator=estimator)
    rfecv.fit(X, y)

    feature_ranks = rfecv.ranking_
    feature_names = X.columns.tolist()

    sorted_features = sorted(zip(feature_names, feature_ranks), key=lambda x: x[1])

    encoded_k = int(rfecv.n_features_)
    encoded_features = [name for name, _ in sorted_features]
    encoded_scores = rfecv.cv_results_['mean_test_score'].tolist()
    encoded_scores_std = rfecv.cv_results_['std_test_score'].tolist()

    decoded_k = 0
    decoded_features = []
    decoded_scores = []
    decoded_scores_std = []

    for i, feature in enumerate(encoded_features):
        decoded_feature = X_inv[feature]
        if decoded_feature in decoded_features:
            if decoded_features[-1] == decoded_feature:
                decoded_scores[-1] = encoded_scores[i]
                decoded_scores_std[-1] = encoded_scores_std[i]
        else:
            decoded_features.append(decoded_feature)
            decoded_scores.append(encoded_scores[i])
            decoded_scores_std.append(encoded_scores_std[i])

        if i == encoded_k - 1:
            decoded_k = len(decoded_features)

    return {
        'valid': True,
        'features': decoded_features,
        'k': decoded_k,
        'kFeatures': decoded_features[:decoded_k],
        'cumulativeScore': decoded_scores,
        'cumulativeScoreStd': decoded_scores_std,
        'encoded': {
            'features': encoded_features,
            'k': encoded_k,
            'kFeatures': encoded_features[:encoded_k],
            'cumulativeScore': encoded_scores,
            'cumulativeScoreStd': encoded_scores_std
        }
    }


# ----------------------------------------------------------------
# SynDiffix Wrapper
# ----------------------------------------------------------------


def get_this_dir():
    script_path = os.path.abspath(__file__)
    return os.path.dirname(script_path)


def get_syndiffix_exe():
    bin_dir = os.environ.get('SYNDIFFIX_BIN')

    if not bin_dir:
        bin_dir = get_this_dir()

    exe_name = 'syndiffix'

    if platform.system() == 'Windows':
        exe_name += '.exe'

    full_path = os.path.join(bin_dir, exe_name)
    if not os.path.exists(full_path):
        print('Could not find executable for SynDiffix. If you want to run from source, add the --dev flag.')
        exit(1)

    return full_path


def run_syndiffix(input_path, output_path, columns, dev=False, extra_args=[], user_args=''):
    if dev:
        run_args = ['dotnet', 'run', '--configuration', 'Release']
        cwd = os.path.join(get_this_dir(), 'src', 'SynDiffix')
    else:
        run_args = [get_syndiffix_exe()]
        cwd = None

    print('Running SynDiffix...')

    columns_args = [c['name'] + ':' + c['type'] for c in columns]

    user_args = shlex.split(user_args)

    syndiffix_process = subprocess.run(
        [*run_args, input_path, '--output', output_path, *extra_args, *user_args, '--columns', *columns_args],
        cwd=cwd
    )

    return syndiffix_process.returncode


def load_csv(path):
    from pandas.errors import ParserError

    df = pd.read_csv(path, keep_default_na=False, na_values=[''], low_memory=False)

    # Try to infer datetime columns.
    for col in df.columns[df.dtypes == 'object']:
        try:
            df[col] = pd.to_datetime(df[col], format='ISO8601')
        except (ParserError, ValueError):
            pass

    return df


def columns_metadata(df):
    columns = []

    for col_name, col_type in zip(df.columns, df.dtypes):
        if pd.api.types.is_bool_dtype(col_type):
            t = 'b'
        elif pd.api.types.is_integer_dtype(col_type):
            t = 'i'
        elif pd.api.types.is_float_dtype(col_type) or pd.api.types.is_numeric_dtype(col_type):
            t = 'r'
        elif pd.api.types.is_datetime64_any_dtype(col_type):
            t = 't'
        elif pd.api.types.is_string_dtype(col_type) or pd.api.types.is_object_dtype(col_type):
            t = 's'
        else:
            raise Exception(f"Unknown type for column '{col_name}'.")

        col_data = {'name': col_name, 'type': t}
        columns.append(col_data)

    return columns


def process_aid_columns(arg):
    if isinstance(arg, list):
        return arg
    elif isinstance(arg, tuple):
        return list(arg)
    elif isinstance(arg, str):
        return [arg]
    else:
        return []


def main(
        input_path: str,
        output_path: str,
        aid_columns: list[str] = [],
        ml_target: str = None,
        ml_features_only: bool = False,
        syndiffix_args: str = '',
        dev: bool = False):
    """
    Runs SynDiffix on CSV file and writes synthetic data to output path.

    Parameters:
        input_path: Path of input CSV file.
        output_path: Path of output CSV file.
        aid_columns: Entity identifier columns. If not specified, assumes one row per entity.
        ml_target: If specified, focuses on this column for better ML prediction.
        ml_features_only: If set, limits columns to only ML features of ml_target.
        syndiffix_args: Extra arguments to pass to syndiffix.
        dev: Compile and run on the fly via 'dotnet run'.
    """

    input_path = os.path.abspath(input_path)
    output_path = os.path.abspath(output_path)

    df = load_csv(input_path)
    columns = columns_metadata(df)

    extra_args = []

    aid_columns = process_aid_columns(aid_columns)
    if len(aid_columns) > 0:
        print(f'AID Columns: {aid_columns}')
        extra_args += ['--aidcolumns', *aid_columns]

    if ml_target:
        print('ML Target: ' + ml_target)

        print('Selecting ML features...')
        features = select_features_ml(df, ml_target)['kFeatures']
        print('ML Features: ' + (', '.join(features)))

        extra_args += [
            '--clustering-maincolumn', ml_target,
            '--clustering-mainfeatures', *features
        ]

        if ml_features_only:
            print('Using only target column and its features.')
            columns = [column for column in columns if column['name'] == ml_target or column['name'] in features]

    return_code = run_syndiffix(input_path, output_path, columns, dev=dev,
                                user_args=syndiffix_args, extra_args=extra_args)

    if return_code == 0:
        print(f"CSV saved to '{output_path}'.")
    else:
        print('Failed to run SynDiffix.')

    exit(return_code)


if __name__ == '__main__':
    fire.Fire(main)
